src/compiler/checker.ts

```diff
@@ -21171,6 +21177,53 @@ namespace ts {
             return isArrayType(type) || !(type.flags & TypeFlags.Nullable) && isTypeAssignableTo(type, anyReadonlyArrayType);
         }
 
+        function tryGetNonShadowedArrayOrTupleType(type: Type) {
+            if (isArrayOrTupleType(type)) {
+                return type;
+            }
+
+            if (!(type.flags & TypeFlags.Intersection)) {
+                return undefined;
+            }
+
+            // pick an intersection constituent if it is an array or tuple type, but only
+            // if any non-array, non-tuple constituents that follow it do not shadow any
+            // tuple-specific members.
+            let arrayOrTuple: TypeReference | undefined;
+            let constituents: Type[] | undefined;
+            for (const constituent of (type as IntersectionType).types) {
+                if (isArrayOrTupleType(constituent)) {
+                    arrayOrTuple = constituent;

```

Comment:
Also this always just picks the last array or tuple type, making it order-dependent which tuple is picked, which seems bad. We should either have tuple-intersecting behavior or return `undefined` when there are multiple arrays/tuples, IMO.

Reply:
We are already order dependent for this case: [Playground link](https://www.typescriptlang.org/play?#code/C4TwDgpgBAglC8UBOECGATA9gOwDYigG0ADKAEgG8BnYJAS2wHMBfYgXQG4AoUSKAIQTI0WPARKUa9JsyjtuXdBADGuVCijKcNKKgBGALlhQAZAO5LV66Fuw69qI4LMwFAejdRsmYFExIlJCglSGwlbF8ACwgUADp4rlsdAA8ARiF9QgAGTigoDyg0o1JJWgYWUjNiUukWOUTtX2SAJiEHbNz8zxbi8moymUq5GvLZYi4uAv9A4IhQ8KiYiHjYnnBoAA0AHgAVKAhk4AgwqmEMHHwoKXLCNgA+IT2Do5Oz0UvCBgAzGKgACQANFAVt9fgB9NhQAD8-ygRmwEAAbjFuLxoHtENs4GZ+HcOF0oDtiiMZPUgA)

Reply:
I don't think intersecting the tuples is viable here, as it introduces far too much complexity. I'm leaning towards what appears to be the current behavior in this case (i.e., picking the right-most), though another viable alternative would be to pick neither if there are more than one (since they would overlap).

Reply:
Picking neither if there's more than one is often what we do in other situations, and order-dependence in inference is bad anywhere it crops up, since it means inference results can shift with file ordering, which is a pain in the butt bug to track down when it's reported later. Our existing left-preferring behavior among multiple matching inferences is already a big "this really should be fixed"; I'd much prefer to bring the cases that introduce order dependence down if possible.